### PR TITLE
Can include model schema definitions in generated $resources

### PR DESCRIPTION
**Pull request by akera-io**
Hi,

we're working on a Kendo data source that can use Strongloop REST services as 'transport' and while things are working just fine for read operations for updates the Kendo data source require some schema information - the model primary key is mandatory else no update operation can take place. 

While one can add that information manually when defining the data source it's a pita not to use the information already available in our models. We've just added an option to include the model definition (properties) when generating the angular resources, this is not included by default so if you think we can merge that back in 'master' we would rather not keep the fork alive. There is a minor change in the client to accommodate this.

Best regards,
Marian Edu

marian.edu@acorn.ro
www.acorn-it.com
www.akera.io

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

**Comment by VR-Architect** [comment]
Can someone please provide an example of how to call the LoopBack model from inside a Kendo UI datasource attrribute?

**Comment by akera-io** [comment]
We have some samples online although using a somehow limited version of the custom Kendo UI data source - http://repository.akera.io:8383/slc/ (Kendo Grid using Akera Angular SDK).

Since then we have extended the datasource and though we can make it generic for all LoopBack models if we can get the model definition exported on the client-side as well... what we do now for akera.io is to have a base model that exposes the 'describe' method and have all models extent that instead.

We're going to release the data source on github, guess we just have to check if the model definition can be retrieved somehow and only include the schema if that is available. Thing is for anything else beside read operations the Kendo datasource need to know the model primary key - for update, delete :(

File: lib/services.template.ejs [hunk 1]
```diff
@@ -252,6 +252,18 @@ module.factory(
...
 
+<% if (meta.modelSchema) { -%>
+    /**
+     * @ngdoc object
+     * @name <%-: moduleName %>.<%- modelName %>#schema
+     * @propertyOf <%-: moduleName %>.<%- modelName %>
+     * @description
+     * The schema of the model represented by this $resource
+     */
+    R.schema = <%- JSON.stringify(meta.modelSchema, null, '\t') -%>
```
Comment (bajtos) [inline lib/services.template.ejs]: Please use 2 spaces for indentation for consistency, we don't use tabs anywhere else.

File: lib/services.js [hunk 2]
```diff
@@ -24,12 +24,21 @@ ejs.filters.q = function(obj) {
...
 
   var models = describeModels(app);
 
+  if (includeSchemas === true) {
+    for (var modelName in models) {
+      var modelProperties = app.models[modelName].definition.rawProperties;
+      models[modelName].modelSchema = {
+        name: modelName,
+        properties: modelProperties
+      }
```
Comment (bajtos) [inline lib/services.js]: Add `;` at the end of the statement.

File: lib/services.template.ejs (hunk 1 above)
Comment (bajtos) [inline lib/services.template.ejs]: Also add `;` at the end of the generated statement.

```
R.schema = <%- ... -%>;
```

**Comment by bajtos** [comment]
Hi @akera-io, thank you for the pull request. Please add a unit-test to verify the implementation and address my minor comments above.

**Comment by bajtos** [comment]
@slnode ok to test

**Comment by akera-io** [comment]
Hi Miroslav,

addressing the comments should be straight forward, for unit-test should we add a new test in services.spec.js or what is your favourite?

Cheers,
Marian Edu

Acorn IT
www.acorn-it.com<http://www.acorn-it.com>
www.akera.io<http://www.akera.io>
+40 740 036 212

On 25 Jan 2016, at 15:27, Miroslav Bajtoš <notifications@github.com<mailto:notifications@github.com>> wrote:


Hi @akera-io<https://github.com/akera-io>, thank you for the pull request. Please add a unit-test to verify the implementation and address my minor comments above.

**Comment by akera-io** [comment]
added unit test, hope this is how we should have added one :)

**Comment by bajtos** [comment]
@akera-io sorry for the delay, this pull request sort of fell through cracks. As I was reviewing the code, I figured out it will be faster for me to write the correct version myself, see #223.

**Pull request closed without merging**